lib/shopify-cli/core/monorail.rb

```diff
@@ -74,42 +72,55 @@ def prompt_for_consent
 
-        def do_post(headers, body, num_retries)
-          Helpers::Async.in_thread do
-            CLI::Kit::Util.begin do
-              Net::HTTP.start(ENDPOINT_URI.host, ENDPOINT_URI.port, use_ssl: ENDPOINT_URI.scheme == 'https') do |http|
-                post = Net::HTTP::Post.new(ENDPOINT_URI.request_uri, headers)
-                post.body = body
-                http.request(post) do |response|
-                  code = response.code.to_i
-                  unless code == 200
-                    raise(MonorailError, "Unexpected status #{code} from Monorail Edge: #{response.body}")
-                  end
-                end
-              end
-            end.retry_after(MonorailError, retries: num_retries) do |error|
-              ShopifyCli::Logger.warn(name) { "#{error.message}; retrying..." }
-            end
-          end
+        def monorail_payload(args:, duration:, result:)
+          {
+            cli_sha: ShopifyCli::Git.sha(dir: ShopifyCli::ROOT),
+            uname: uname,
+            args: args,
+            timestamp: Time.now.utc.iso8601(MICROSECOND_PRECISION),
+            duration: duration,
+            result: result,
+          }
         end
 
-        def enabled?
-          # we only want to send Monorail events in production or when explicitly developing
-          Context.new.system? || ENV['MONORAIL_REAL_EVENTS'] == '1'
+        def uname
+          @uname ||= %x{uname -a}
         end
 
-        def consented?
-          ShopifyCli::Config.get_bool('analytics', 'enabled')
+        def ruby_version
+          @ruby_version ||= RUBY_VERSION
         end
 
-        def events
-          @events ||= ShopifyCli::Log.new(ShopifyCli::EVENTS_FILE, mode: 'a+')
-        end
+        def send_event(event)
+          return unless enabled?
+          return unless consented?
 
-        def mtime
-          @mtime ||= Time.parse(File.read(Helpers::FS.ensure_file(ShopifyCli::EVENTS_MTIME)))
-        rescue ArgumentError
-          Time.new(2019)
+          headers = {
+            'Content-Type': 'application/json; charset=utf-8',
+            'X-Monorail-Edge-Event-Created-At-Ms': Time.parse(event[:payload][:timestamp]).to_i.to_s,
+            'X-Monorail-Edge-Event-Sent-At-Ms': Time.now.utc.to_i.to_s,
+          }
+
+          begin
+            Net::HTTP.start(
+              ENDPOINT_URI.host,
+              ENDPOINT_URI.port,
+              # timeouts for opening a connection, reading, writing (in seconds)
+              open_timeout: 0.2, read_timeout: 0.2, write_timeout: 0.2,
+              use_ssl: ENDPOINT_URI.scheme == 'https'
+            ) do |http|
+              post = Net::HTTP::Post.new(ENDPOINT_URI.request_uri, headers)
+              post.body = JSON.dump(event)
+              http.request(post) do |response|
+                code = response.code.to_i
+                unless code == 200
+                  raise(MonorailError, "Unexpected status #{code} from Monorail Edge: #{response.body}")
```

Comment:
I know you didn't do that but maybe we should only do this if the CLI is in debug and fail silently otherwise?

Reply:
Ya good point, right now I think it would just get caught by the `rescue` below, but we probably don't need to `raise` only to catch, this could just as easily be a `ShopifyCli::Logger.error` since I don't think we display those to the user.

Reply:
My bad, logs DO show to the user so I'm getting rid of this entirely and just failing silently instead.